lib/ansible/plugins/connection/nspawn.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import pipes
+import subprocess
+import traceback
+import shlex
+
+from ansible import constants as C
+from ansible.errors import AnsibleError
+from ansible.plugins.connection import ConnectionBase
+from ansible.module_utils.basic import is_executable
+from ansible.utils.unicode import to_bytes
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+BUFSIZE = 65536
+
+
+class Connection(ConnectionBase):
+    ''' Local nspawn based connections '''
+
+    transport = 'nspawn'
+    has_pipelining = True
+    become_methods = frozenset(C.BECOME_METHODS)
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin,
+                *args, **kwargs)
+
+        display.vvv("NSPAWN ARGS %s" % self._play_context.nspawn_args)
+
+        self.chroot = self._play_context.remote_addr
+
+        if os.geteuid() != 0:
+            raise AnsibleError("nspawn connection requires running as root")
+
+        # we're running as root on the local system so do some
+        # trivial checks for ensuring 'host' is actually a chroot'able dir
+        if not os.path.isdir(self.chroot):
+            raise AnsibleError("%s is not a directory" % self.chroot)
+
+        chrootsh = os.path.join(self.chroot, 'bin/sh')
+        if not is_executable(chrootsh):
+            raise AnsibleError("%s does not look like a chrootable dir (/bin/sh missing)" % self.chroot)
+
+        self.nspawn_cmd = 'systemd-nspawn'
+
+    def _connect(self):
+      pass
+
+    def _buffered_exec_command(self, cmd, stdin=subprocess.PIPE):
+        ''' run a command on the chroot.  This is only needed for
+        implementing put_file() get_file() so that we don't have to
+        read the whole file into memory.
+
+        compared to exec_command() it looses some niceties like being
+        able to return the process's exit code immediately.
+        '''
+        executable = (
+                C.DEFAULT_EXECUTABLE.split()[0]
+                if C.DEFAULT_EXECUTABLE
+                else '/bin/sh')
+
+        nspawn_args = shlex.split(self._play_context.nspawn_args)
```

Comment:
Unfortunately, shlex.split() needs a bit of code to make it compatible with both python-2.6 and python3.x

On python-2.6, it only works on byte strings.  In python3.x it only works with text strings.  So you have to test for python version and then convert appropriately.  Code like the following is what I use:

``` python
from ansible.compat import six
from ansible.module_utils._text import to_bytes, to_text
[...]
nspawn_args = self._play_context.nspawn_args
if six.PY2:
    nspawn_args = shlex.split(to_bytes(nspawn_args, errors='surrogate_or_strict'))
else:
    nspawn_args = shlex.split(to_text(nspawn_args, errors='surrogate_or_strict'))